### PR TITLE
Read with timeout bug fix

### PR DESCRIPTION
**Pull request by KeonF-ross**
## 📖 Description

Encountered some sort of memory bug while working on the windows build. Seemed to be a use after free occurring when the REST read operation is timed out. Made changes to how objects are passed between the `SocketReader` and the `read_with_timeout` coroutine to avoid lifetime issues. Also modified error handling so that the coroutine itself no longer throws but returns an error code. Beforehand, the throws from the coroutine could be caught by the debugger which was annoying, and the change to returning error codes works nicely with the memory fix. Changes were also made and tested on the windows build.

I also found a separate bug as well while fixing it. The `catena::StatusCode::DEADLINE_EXCEEDED` was missing a mapping to a HTTP status code. This would cause the `SocketWriter` to fail and the socket to be hung up instead of returning an error message.

---

## ✅ Definition of Done (DoD)
- [x] Feature is implemented and works as expected
- [x] Edge cases are handled
- [x] Tests are added or updated (if applicable)
- [ ] Code is reviewed and approved

---

## 🛠️ Implementation Notes

- Modified `ServiceImpl.cpp`
   - The `tcp::socket` is now created as a shared pointer which is moved onto the worker thread.
   - This is done to support changes to `read_with_timeout`.
- Modified `SocketReader.cpp`
   -  Added `ReadResult` struct which contains output of `read_with_timeout`.
   -  `read_with_timeout` now returns error code instead of throwing.
   -  `read_with_timeout` creates and returns the buffer.
- Modified `SocketReader_test.cpp`
   - Updated to use shared pointer to `tcp::socket`.
- Modified `SocketWriter.h`
   - `catena::StatusCode::DEADLINE_EXCEEDED` now maps to `408`, `Request Timeout`.
---

## 🔗 Related Issues / Links

---
## 📝 Additional Context

If you have any ideas for improving/simplifying my fix feel free to share.

---
## 🚧 Risks / Open Questions

---
## ✨ Updates
<!-- (Optional) List summarizing changes made when editing the pull request>

### CODE REVIEW

**Review by abrown-RV — CHANGES_REQUESTED** [review]
if possible can you add a unit test that forces the bad state to occur? basically it would fail without the change, but now it should always pass.

**Comment by KeonF-ross** [comment]
> if possible can you add a unit test that forces the bad state to occur? basically it would fail without the change, but now it should always pass.

The MissingEndHeaders, MultipleIncorrect and InvalidContentLength would all cause it to fail in the windows build when they timed out. This is the error that you would get:
571: minkernel\crts\ucrt\src\appcrt\heap\debug_heap.cpp(904) : Assertion failed: _CrtIsValidHeapPointer(block)
571: minkernel\crts\ucrt\src\appcrt\heap\debug_heap.cpp(908) : Assertion failed: is_block_type_valid(header->_block_use)

Did you want me to try adding a test case that would make the linux build (pre-changes) to fail as well?

**Comment by KeonF-ross** [comment]
> if possible can you add a unit test that forces the bad state to occur? basically it would fail without the change, but now it should always pass.

So I tried making one but there doesn't seem to be a way to make a unit test on Linux that can force the bad state and catch the error. Instead, I've done a bit of manual investigating using Valgrind (more specifically Helgrind) to check if the error is being fixed. The command used was `valgrind --tool=helgrind --log-file=vg.log ./unittests/REST/REST_SocketReader_test --gtest_filter='RESTSocketReaderTests.SocketReader_MissingEndHeaders'`. This runs the MissingEndHeaders test case which is guaranteed to cause a timeout on the initial `read_until_with_timeout()`. Below is the results of running this test case with Helgrind.

image

```
==44518== Possible data race during read of size 8 at 0x1FFEFFF318 by thread #1
==44518== Locks held: none
==44518==    at 0x1EA9C2: std::_Vector_base<char, std::allocator<char> >::~_Vector_base() (stl_vector.h:370)
==44518==    by 0x1E72AA: std::vector<char, std::allocator<char> >::~vector() (stl_vector.h:738)
==44518==    by 0x1E61BD: boost::asio::basic_streambuf<std::allocator<char> >::~basic_streambuf() (basic_streambuf.hpp:111)
==44518==    by 0x1DED98: catena::REST::SocketReader::read(boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::any_io_executor>&, unsigned int) (SocketReader.cpp:282)
==44518==    by 0x1715C2: RESTSocketReaderTests_SocketReader_MissingEndHeaders_Test::TestBody() (SocketReader_test.cpp:739)
==44518==    by 0x24436E: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (in /home/keonfoster/Catena/build/cpp/unittests/REST/REST_SocketReader_test)
==44518==    by 0x22B6B5: testing::Test::Run() (in /home/keonfoster/Catena/build/cpp/unittests/REST/REST_SocketReader_test)
==44518==    by 0x22B874: testing::TestInfo::Run() (in /home/keonfoster/Catena/build/cpp/unittests/REST/REST_SocketReader_test)
==44518==    by 0x22BA5E: testing::TestSuite::Run() (in /home/keonfoster/Catena/build/cpp/unittests/REST/REST_SocketReader_test)
==44518==    by 0x2398CB: testing::internal::UnitTestImpl::RunAllTests() (in /home/keonfoster/Catena/build/cpp/unittests/REST/REST_SocketReader_test)
==44518==    by 0x244A46: bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (in /home/keonfoster/Catena/build/cpp/unittests/REST/REST_SocketReader_test)
==44518==    by 0x22BC57: testing::UnitTest::Run() (in /home/keonfoster/Catena/build/cpp/unittests/REST/REST_SocketReader_test)
==44518== 
==44518== This conflicts with a previous write of size 8 by thread #2
==44518== Locks held: none
==44518==    at 0x1EFFA3: std::vector<char, std::allocator<char> >::_M_default_append(unsigned long) (vector.tcc:707)
==44518==    by 0x1ECE4A: std::vector<char, std::allocator<char> >::resize(unsigned long) (stl_vector.h:1016)
==44518==    by 0x1FB322: boost::asio::basic_streambuf<std::allocator<char> >::reserve(unsigned long) (basic_streambuf.hpp:335)
==44518==    by 0x1F9455: boost::asio::basic_streambuf<std::allocator<char> >::prepare(unsigned long) (basic_streambuf.hpp:217)
==44518==    by 0x1F74C7: boost::asio::basic_streambuf_ref<std::allocator<char> >::prepare(unsigned long) (basic_streambuf.hpp:428)
==44518==    by 0x1F506F: boost::asio::detail::read_until_delim_string_op_v1<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::any_io_executor>, boost::asio::basic_streambuf_ref<std::allocator<char> >, boost::asio::detail::awaitable_handler<boost::asio::any_io_executor, boost::system::error_code, unsigned long> >::operator()(boost::system::error_code, unsigned long, int) (read_until.hpp:1227)
==44518==    by 0x2058CA: boost::asio::detail::binder2<boost::asio::detail::read_until_delim_string_op_v1<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::any_io_executor>, boost::asio::basic_streambuf_ref<std::allocator<char> >, boost::asio::detail::awaitable_handler<boost::asio::any_io_executor, boost::system::error_code, unsigned long> >, boost::system::error_code, unsigned long>::operator()() (bind_handler.hpp:289)
==44518==    by 0x204E2B: void boost::asio::asio_handler_invoke<boost::asio::detail::binder2<boost::asio::detail::read_until_delim_string_op_v1<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::any_io_executor>, boost::asio::basic_streambuf_ref<std::allocator<char> >, boost::asio::detail::awaitable_handler<boost::asio::any_io_executor, boost::system::error_code, unsigned long> >, boost::system::error_code, unsigned long> >(boost::asio::detail::binder2<boost::asio::detail::read_until_delim_string_op_v1<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::any_io_executor>, boost::asio::basic_streambuf_ref<std::allocator<char> >, boost::asio::detail::awaitable_handler<boost::asio::any_io_executor, boost::system::error_code, unsigned long> >, boost::system::error_code, unsigned long>&, ...) (handler_invoke_hook.hpp:88)
==44518==  Address 0x1ffefff318 is on thread #1's stack
==44518==  in frame #3, created by catena::REST::SocketReader::read(boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::any_io_executor>&, unsigned int) (SocketReader.cpp:93)
```

The above error is the type of error corresponding to the bug this PR aims to fix. This points to the main cause of the errors on the windows build: A buffer lifetime bug. The buffer is created on `SocketReader::read()`'s stack and is passed into the `async_read_until()` which is running on the `runner` thread of `ServiceImpl::run()`. When the timeout occurs, `SocketReader::read()` throws and triggers the local buffer to be destroyed, however, for whatever reason the `async_read_until()` is still touching the buffer, causing the error. 

My fix for this works by allocating the buffer on the heap as part of the `read_until_with_timeout()` and then moving it to `SocketReader::read()` if and only if the `async_read_until()` completed successfully. So, in the case of a timeout, the buffer is never touched by `SocketReader::read()` and is never destroyed early. Below is the Helgrind result of my fix.

image

As you can see, a lot of the errors disappeared, and the specific type that I was looking for was no longer in the results. The remaining errors all seem to be noise picked up by Helgrind.

The same fix was applied to `read_with_timeout()`, however its buffer works a bit differently so Helgrind doesn't seem to be able to properly pick up the errors so I can't do a deeper check of it, but it does pass on the windows build as well.

As part of the PR, I also changed the socket to use a shared pointer. This might be unnecessary, but I'd rather be safe than sorry.

File: sdks/cpp/connections/REST/src/SocketReader.cpp [hunk 1]
```diff
@@ -28,35 +28,68 @@ static inline bool iequals_header_name(std::string_view a, std::string_view b) {
...
+    using namespace boost::asio;
+    using namespace experimental::awaitable_operators;
+    auto buf = std::make_unique<boost::asio::streambuf>();
+    steady_timer timer(socket->get_executor(), std::chrono::milliseconds(timeout));
+
 
     // Async_read and timer run simultaneously, the other cancels when one finishes
-    auto result = co_await (boost::asio::async_read_until(socket, buf, delim, boost::asio::use_awaitable) ||
-        timer.async_wait(boost::asio::use_awaitable)
+    auto result = co_await (async_read_until(*socket, *buf, "\r\n\r\n", as_tuple(use_awaitable)) ||
```
Comment (abrown-RV) [inline sdks/cpp/connections/REST/src/SocketReader.cpp]: I feel like delim should go back? or the func should be renamed to reflect what its reading until

File: sdks/cpp/connections/REST/src/SocketReader.cpp [hunk 2]
```diff
@@ -112,18 +145,14 @@ void SocketReader::read(tcp::socket& socket, uint32_t timeout) {
...
-        result.get();
-    } catch(const catena::exception_with_status& e) {
-        throw;
-    } catch(...) {
+    auto fut = boost::asio::co_spawn(socket->get_executor(), read_until_with_timeout(socket, timeout), boost::asio::use_future);
+    ReadResult result = fut.get();
+    if (result.ec == boost::asio::error::timed_out) {
+        throw catena::exception_with_status("Timed out", catena::StatusCode::DEADLINE_EXCEEDED);
+    } else if (result.ec) {
         throw catena::exception_with_status("Read error", catena::StatusCode::UNKNOWN);
```
Comment (abrown-RV) [inline sdks/cpp/connections/REST/src/SocketReader.cpp:153]: Is there an easy way to hit this, and other branches like it? I guess async_read would need to return some kind of error. Maybe if the client hangs up unexpectedly or something?

**Review by abrown-RV — CHANGES_REQUESTED** [review]
very cool, nice work with valgrind

**Review by abrown-RV — APPROVED** [review]
